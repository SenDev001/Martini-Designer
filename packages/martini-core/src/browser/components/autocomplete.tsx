import * as React from "react";
import styled from "styled-components";
import { ElementExt } from '@phosphor/domutils';

export interface Suggestion {
    label: string;
    iconClass?: string;
}

export interface AutocompleteProps {
    value: string;
    suggestions: Suggestion[];
    onInputChange: (input: string) => void;
    onInputBlur?: (e: React.FocusEvent<HTMLInputElement>, input: string) => void;
    onSuggestionSelect?: (input: string) => void;
    onInputKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
    onInputKeyUp?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
    inputStyle?: React.CSSProperties;
    autoFocus?: boolean;
}

const UlStyles = styled.ul`
    color: var(--theia-input-foreground);
    background-color: var(--theia-input-background);
    border: 1px solid var(--theia-inputOption-activeBorder);
    border-top: 0;
    position: absolute;
    top: 25px;
    left: 0px;
    width: calc(100% - 2px);
    list-style: none;
    margin-top: 0;
    max-height: 143px;
    overflow-y: auto;
    padding-left: 0;
    z-index: 20;

    li {
        padding: 4px;
    }

    .suggestion-active, li:hover {
        background-color: var(--theia-list-activeSelectionBackground);
        color: var(--theia-input-foreground);
        cursor: pointer;
    }
`;

export const Autocomplete: React.FC<AutocompleteProps> = ({
    value,
    suggestions,
    onInputChange,
    onInputBlur,
    onSuggestionSelect,
    onInputKeyDown,
    onInputKeyUp,
    inputStyle,
    autoFocus
}) => {
    const [activeSuggestion, setActiveSuggestion] = React.useState(0);
    const [filteredSuggestions, setFilteredSuggestions] = React.useState([] as Suggestion[]);
    const [showSuggestions, setShowSuggestions] = React.useState(false);
    const mounted = React.useRef(true);
    React.useEffect(() => {
        if (autoFocus && inputRef.current)
            inputRef.current.setSelectionRange(0, value.length);
        return () => { mounted.current = false; };
    }, []);
    const inputRef = React.useRef<HTMLInputElement>(null);
    const ulRef = React.useRef<HTMLUListElement>(null);

    const filter = (input: string) => input.length === 0 ? suggestions : suggestions.filter(
        suggestion =>
            suggestion.label.toLowerCase().includes(input)
    );

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const input = e.target.value;
        setActiveSuggestion(0);
        setFilteredSuggestions(filter(input.toLowerCase()));
        setShowSuggestions(true);
        onInputChange(input);
    };

    const onClick = (suggestion: Suggestion) => {
        setActiveSuggestion(0);
        setFilteredSuggestions([]);
        setShowSuggestions(false);
        onInputChange(suggestion.label);
        if (onSuggestionSelect)
            onSuggestionSelect(suggestion.label);
        inputRef.current?.focus();
    };

    const onKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === "Enter" && showSuggestions && filteredSuggestions.length !== 0) {
            const label = filteredSuggestions[activeSuggestion].label;
            setActiveSuggestion(0);
            setShowSuggestions(false);
            onInputChange(label);
            if (onSuggestionSelect)
                onSuggestionSelect(label);
            return;
        }
        else if (e.key === "Escape" && showSuggestions) {
            setShowSuggestions(false);
            return;
        }
        else if (e.key === "ArrowUp") {
            e.preventDefault();
            let next = activeSuggestion - 1;

            if (next < 0)
                next = filteredSuggestions.length - 1;

            setActiveSuggestion(next);
            if (ulRef.current)
                ElementExt.scrollIntoViewIfNeeded(ulRef.current, ulRef.current.childNodes[next] as HTMLLIElement);
            return;
        }
        else if (e.key === "ArrowDown") {
            e.preventDefault();
            let next = activeSuggestion + 1;

            if (next >= filteredSuggestions.length)
                next = 0;

            setActiveSuggestion(next);
            if (ulRef.current)
                ElementExt.scrollIntoViewIfNeeded(ulRef.current, ulRef.current.childNodes[next] as HTMLLIElement);
            return;
        }
        else if (!showSuggestions && e.key === " " && e.ctrlKey) {
            setShowSuggestions(true);
            return;
        }

        if (onInputKeyUp)
            onInputKeyUp(e);
    };

    const onBlur = (e: React.FocusEvent<HTMLInputElement>) => {
        // when e.relatedTarget === null it means it probably click one of the suggestions
        if (e.relatedTarget !== null && onInputBlur) {
            onInputBlur(e, value);
            setShowSuggestions(false);
        }
    };

    let suggestionsListComponent;

    if (showSuggestions && filteredSuggestions.length) {
        suggestionsListComponent = (
            <UlStyles ref={ulRef}>
                {filteredSuggestions.map((suggestion, index) => {
                    return (
                        <li
                            className={index === activeSuggestion ? "suggestion-active" : undefined}
                            key={suggestion.label}
                            onClick={() => onClick(suggestion)}
                        >
                            {suggestion.label}
                        </li>
                    );
                })}
            </UlStyles>
        );
    }

    return <>
        <input
            type="text"
            autoFocus={autoFocus}
            ref={inputRef}
            onChange={onChange}
            onBlur={onBlur}
            value={value}
            onKeyDown={onInputKeyDown}
            onKeyUp={onKeyUp}
            className="theia-input"
            style={inputStyle}
        />
        {suggestionsListComponent}
    </>;
};
