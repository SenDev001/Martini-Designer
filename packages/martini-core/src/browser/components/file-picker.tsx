import * as React from "react";

export interface FilePickerProps {
    extensions?: string;
    placeholder?: string;
    value?: File | string;
    onChange?: (fileOrUrl: File | string) => void;
}

export const FilePicker: React.FC<FilePickerProps> = ({extensions, placeholder, value, onChange, children}) => {
    const [file, setFile] = React.useState(!value ? "" : value instanceof File ? value.name : value);
    const fileInput = React.useRef<HTMLInputElement>(null);
    const [invalidDrop, setInvalidDrop] = React.useState(false);

    const isValidFile = (file: File) => {
        if (!extensions)
            return true;
        return extensions.split(/,/).some(ext => file.name.endsWith(ext));
    };
    const handleFiles = (files: FileList) => {
        if (files && files.length !== 0) {
            const first = files.item(0);
            if (first && isValidFile(first)) {
                setFile(first.name);
                if (onChange)
                    onChange(first);
                return true;
            }
        }
    };
    const handleFileInputChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        const files = e.target.files;
        if (files)
            handleFiles(files);
    };
    const handleTextInputChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        setFile(e.target.value);
        onChange!(e.target.value);
    };

    const handleDrop = (e: React.DragEvent) => {
        if (!handleFiles(e.dataTransfer.files)) {
            e.dataTransfer.dropEffect = "none";
            e.dataTransfer.effectAllowed = "none";
            setInvalidDrop(true);
            setTimeout(() => setInvalidDrop(false), 3000);
        }
    };

    return <div
        style={{display: "grid", gridTemplateColumns: "1fr max-content"}}
        onDrop={handleDrop}>
        <input type="file"
               style={{display: "none"}}
               ref={fileInput}
               accept={extensions}
               onChange={handleFileInputChanged}/>
        <input type="text"
               className="theia-input"
               placeholder={placeholder}
               onChange={handleTextInputChanged}
               style={{border: !invalidDrop ? "" : "var(--theia-border-width) solid var(--theia-errorForeground)"}}
               value={file}/>
        <input type="button"
               className="theia-button"
               value="Browse..."
               onClick={() => fileInput.current && fileInput.current.click()}/>
        {children}
    </div>;
};
