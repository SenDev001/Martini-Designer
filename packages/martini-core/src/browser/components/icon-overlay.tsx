import * as React from "react";

export interface IconProps {
    iconClass: string;
}

export const Icon: React.FC<IconProps> = ({ iconClass, children }) => (<div
    className={`martini-tree-icon ${iconClass}`}
    style={{ position: "relative", marginRight: "4px" }}
>
    {children}
</div>);

export interface OverlayIconProps extends IconProps {
    position: "top-left" | "top-right" | "bottom-left" | "bottom-right";
    width?: number;
    height?: number;
}

export const OverlayIcon: React.FC<OverlayIconProps> = (props) => {
    return <div className={props.iconClass} style={getStyle(props)} />;
};

const getStyle = (props: OverlayIconProps) => {
    const width = props.width === undefined ? 8 : props.width;
    const height = props.height === undefined ? 8 : props.height;
    const style: any = {
        backgroundSize: `${width}px ${height}px`,
        backgroundPosition: "center center",
        backgroundRepeat: "no-repeat",
        position: "absolute",
        width,
        height
    };

    switch (props.position) {
        case "top-left":
            style.left = "0";
            style.top = "0";
            break;
        case "top-right":
            style.right = "0";
            style.top = "0";
            break;
        case "bottom-left":
            style.left = "0";
            style.bottom = "0";
            break;
        case "bottom-right":
            style.right = "0";
            style.bottom = "0";
            break;
    }
    return style;
};
