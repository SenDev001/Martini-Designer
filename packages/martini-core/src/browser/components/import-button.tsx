import { CommandRegistry } from "@phosphor/commands";
import { Menu } from "@phosphor/widgets";
import { SingleTextInputDialog } from "@theia/core/lib/browser";
import * as React from "react";
import { isValidUrl } from "../../common/fs/file-util";

export interface ImportButtonProps {
    onFileImport?: (file: File) => void;
    onUrlImport?: (url: string) => void;
    extensions?: string;
}

export const ImportButton: React.FC<ImportButtonProps> = ({ onFileImport, onUrlImport, extensions }) => {
    const fileInputRef = React.useRef<HTMLInputElement>(null);
    const handleClick = (e: React.MouseEvent<HTMLInputElement>) => {
        const commands = new CommandRegistry();
        if (onFileImport) {
            commands.addCommand("fromFile", {
                label: "From File",
                execute: () => {
                    fileInputRef.current?.click();
                }
            });
        }

        if (onUrlImport) {
            commands.addCommand("fromUrl", {
                label: "From URL",
                execute: async () => {
                    const dlg = new SingleTextInputDialog({
                        title: "Enter a URL",
                        confirmButtonLabel: "Import",
                        validate: (url: string) => {
                            if (!isValidUrl(url))
                                return "Not a valid URL.";
                            return true;
                        }
                    });
                    const result = await dlg.open();
                    if (result) onUrlImport(result);
                }
            });
        }
        const menu = new Menu({
            commands
        });
        if (onFileImport)
            menu.addItem({
                command: "fromFile"
            });
        if (onUrlImport)
            menu.addItem({
                command: "fromUrl"
            });
        const bounds = e.currentTarget.getBoundingClientRect();
        menu.open(bounds.left, bounds.bottom);
    };

    const handleFileInput = (e: React.ChangeEvent<HTMLInputElement>) => {
        const files = e.target.files;
        if (files && files.length !== 0) {
            const first = files.item(0);
            if (first) onFileImport!(first);
        }
    };

    return <div>
        <input
            type="file"
            ref={fileInputRef}
            style={{ display: "none" }}
            accept={extensions}
            onChange={handleFileInput}
        />
        <input
            type="button"
            className="theia-button"
            style={{ marginLeft: 0 }}
            value="Import..."
            onClick={handleClick} />
    </div>;
};
