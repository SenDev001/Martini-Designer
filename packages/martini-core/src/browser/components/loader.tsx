import * as React from "react";
import styled from "styled-components";

const LoaderStyles = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;

    .wrapper {
        display: grid;
        width: 100%;
        height: 100%;
        justify-items: center;
        grid-row-gap: 10px;
    }

    .progress-loader-end {
        align-self: end;
    }

    .progress-loader-center {
        align-self: center;
    }

    .progress-message {
        align-self: start;
    }
`;

export interface LoaderProps {
    message?: string;
    style?: React.CSSProperties;
}

export const Loader: React.FC<LoaderProps> = ({ message, style }) => (<LoaderStyles style={style}>
    <div className="wrapper">
        <div className={message ? "loader-large progress-loader-end" : "loader-large progress-loader-center"} />
        {message && <div className="progress-message">{message}</div>}
    </div>
</LoaderStyles>);
