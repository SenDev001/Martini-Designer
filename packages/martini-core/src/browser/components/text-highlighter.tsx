import * as React from "react";

export interface TextHighlighterProps {
    /**
     * Boolean which determines if the search should be
     * case sensitive or not. Default is false
     */
    isCaseSensitive?: boolean;
    /**
     * Boolean which determines if the search is global
     * or stops at first occurence. Default is true
     */
    isGlobalSearch?: boolean;
    /**
     * Custom css class for highlighted nodes
     */
    highlightClass?: string;
    /**
     * Custom css class for non highlighted nodes
     */
    className?: string;
    /**
     * Custom css style for highlighted nodes. Default is
     *  `font-weight: bold`
     */
    highlightStyle?: React.CSSProperties;
    /**
     * custom wrap element for non highlighted texts. Default is
     * `span`
     */
    wrapElement?: string;
    /**
     * custom wrap element for highlighted texts. Default is
     * `span`
     */
    highlightWrapElement?: string;

    /**
     * Search string or regex or an array of strings
     */
    search?: string | RegExp | string[];
}

/**
 * This component is used to highlight a search pattern in its children.
 * Port of `react-highlight-textinput`.
 * See https://www.npmjs.com/package/react-highlight-textinput
 */
export class TextHighlighter extends React.Component<TextHighlighterProps> {
    static readonly defaultProps: TextHighlighterProps = {
        isCaseSensitive: false,
        wrapElement: "span",
        highlightWrapElement: "span",
        highlightStyle: { backgroundColor: "var(--theia-list-filterMatchBackground)" },
        isGlobalSearch: true
    };

    /**
     * Renders given text into a span element
     * @param str input string expression to render plain element
     * @param key unique key for the element
     */
    private renderNormal(str: string, key: any): React.ReactNode {
        const props = {
            key,
            className: this.props.className ? this.props.className : null
        };
        const children = str;
        return React.createElement(this.props.wrapElement!, props, children);
    }

    /**
     * Renders given text into a highlighted span element
     * @param str input string expression to render highlighted element
     * @param key unique key for the element
     */
    private renderHighlighted(str: string, key: any) {
        const { highlightClass, highlightStyle } = this.props;
        const props = {
            key,
            className: highlightClass ? highlightClass : null,
            style: !highlightClass ? highlightStyle : {}
        };
        const children = str;
        return React.createElement(this.props.highlightWrapElement!, props, children);

    }

    /**
    * Searches for given expression in a string and render it into a collection of
    * highlighted and plain elements
    * @param text  string expression to search in
    * @param search regular expression to search for
    */
    private highlight(text: string, search: RegExp) {
        const resultHTML = [];
        let key = 0;
        if (!this.props.isGlobalSearch) {
            const matchedArr = search.exec(text);
            if (!matchedArr) {
                resultHTML.push(this.renderNormal(text, key));
                return resultHTML;
            } else {
                const begin = matchedArr.index, end = matchedArr.index + matchedArr[0].length;

                const normal = text.slice(0, begin), highlight = text.slice(begin, end);
                if (normal) { resultHTML.push(this.renderNormal(normal, key)); key++; }
                if (highlight) { resultHTML.push(this.renderHighlighted(highlight, key)); key++; }
                resultHTML.push(this.renderNormal(text.slice(end), key));

            }

            return resultHTML;
        }

        while (text) {
            const matchedArr = search.exec(text);
            if (!matchedArr) {
                resultHTML.push(this.renderNormal(text, key));
                return resultHTML;
            } else {
                const begin = matchedArr.index, end = matchedArr.index + matchedArr[0].length;
                if (begin === 0 && end === 0) {
                    return resultHTML;
                }
                const normal = text.slice(0, begin), highlight = text.slice(begin, end);
                if (normal) { resultHTML.push(this.renderNormal(normal, key)); key++; }
                if (highlight) { resultHTML.push(this.renderHighlighted(highlight, key)); key++; }

                text = text.slice(end);
            }


        }

        return resultHTML;
    }

    /**
     * Recursively parses the html inside parent node and highlights the passed string
     * @param children valid object of children elements of parent node
     * @param search regular expression to search for
     */
    private parseAndHighlight(children: React.ReactElement, search: RegExp): any {
        if (!children) {
            return children;
        }
        return React.Children.map(children, (child, idx) => {

            if (this.isChildrenPrimitive(child)) {
                return this.highlight(child as unknown as string, search);
            } else {
                const chld = child || <React.Fragment />;
                const childOfChild = chld.props && chld.props.children;
                const parsedChildren = this.parseAndHighlight(childOfChild, search);
                return React.cloneElement(chld, chld.props, parsedChildren);
            }
        });

    }

    /**
     * Determines if the children passed are primitives or not
     * @param children child elements
     */
    private isChildrenPrimitive(children: React.ReactNode): boolean {
        return (/string|number|boolean/).test(typeof children);
    }

    /**
     * Get the escaped version of given string
     * @param str Input string for converting into escaped string
     * @return Escaped string
     */
    private escapeString(str: string) {
        return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    }

    /**
    * Convert a given seacrh string in the form of a regular expression.
    * @param search search string for converting into RegExp Object
    * @param isEscaped detemines whether the search string is escaped or not
    * @return output RegExp object
    */
    private getSearchRegex(search: string | RegExp, isEscaped: boolean): RegExp {
        if (search instanceof RegExp) {
            return search;
        }
        const escapedSearch = isEscaped ? search : this.escapeString(search);
        const caseSense = this.props.isCaseSensitive ? '' : 'i';
        return new RegExp(escapedSearch, caseSense);
    }


    render() {
        const { search, children } = this.props;
        if (!search) {
            return children;
        }
        let sRegex;
        /**If search prop is an Array then escape each value */
        if (search instanceof Array) {
            const finalStr = search.map(search => this.escapeString(search)).join("|");
            sRegex = this.getSearchRegex(finalStr, true);
        } else {
            sRegex = this.getSearchRegex(search, false);
        }

        /**
         * If childdren of TextHighlighter are not text elements then
         * parse the nodes first
         */
        if (!this.isChildrenPrimitive(children)) {
            if (React.Children.count(children) > 1) {
                console.error("TextHighlighter must contain one wrapper element as a child");
                return;
            } else if (!React.isValidElement(children)) {
                console.error("TextHighlighter does not contain valid react elements");
                return;
            }

            /**the immediate child of TextHighlighter will be considered parent for parsed elements**/
            const parentNode = children;
            /**Parse and highlight search words in children */
            const parsedChildren = this.parseAndHighlight(parentNode.props.children, sRegex);
            /**Clone the parent element */
            const parsedParentNode = React.cloneElement(parentNode, parentNode.props, parsedChildren);
            return (
                <React.Fragment>
                    {
                        parsedParentNode
                    }
                </React.Fragment>

            );

        }

        return <span className="thWrapper">
            {
                this.highlight(this.props.children as string, sRegex)
            }
        </span>;
    }
}
