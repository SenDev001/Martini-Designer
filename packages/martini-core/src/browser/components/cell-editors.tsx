import { MaybePromise } from "@theia/core";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import styled from "styled-components";
import * as yup from "yup";
import { Autocomplete, Suggestion } from "./autocomplete";

export interface CellEditor<T> {
    value: T;
    validate?: (value: T) => string | undefined;
    onApply?: (value: T) => void;
    onCancel?: () => void;
    readOnly: boolean;
}

export const CellEditorProvider = Symbol("CellEditorProvider");

export interface CellEditorProvider<T> {
    (props: CellEditorProps<T>): React.ReactNode;
}

export interface CellEditorProps<T> {
    editor: CellEditor<T>;
}

export interface TextCellEditorProps extends CellEditorProps<string> {
    minLength?: number;
    maxLength?: number;
    suggestions?: Suggestion[];
    autoFocus?: boolean;
}

const TextCellEditorStyles = styled.div`
    width: 100%;
    display: grid;
    margin: 0;
    padding: 0;
    overflow-y: visible;
    position: relative;

    input {
        min-width: 0;
    }
`;

export const TextCellEditor: React.FC<TextCellEditorProps> = ({
    editor,
    minLength,
    maxLength,
    suggestions,
    autoFocus
}) => {
    const schema = React.useMemo(() => yup.string().min(minLength || 0).max(maxLength || Number.MAX_VALUE), []);
    const getValue = () =>
        editor.value === undefined ? "" : editor.value;
    const valueChanged = React.useRef(false);
    const [value, setValue] = React.useState(getValue());
    const [error, setError] = React.useState("");

    const validate = (input: string): string | undefined => {
        try {
            schema.validateSync(input);
        } catch (error) {
            return error.errors[0];
        }

        if (editor.validate)
            return editor.validate(input);
    };

    const onChange = (input: string) => {
        if (editor.readOnly)
            return;
        valueChanged.current = true;
        setValue(input);
        setError(validate(input) || "");
    };

    const apply = (input: string): boolean => {
        if (validate(input) !== undefined) return false;
        if (editor.value !== input && editor.onApply) {
            editor.onApply(input);
            return true;
        }

        if (editor.onCancel)
            editor.onCancel();
        return false;
    };

    const handleKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === "Enter") {
            e.preventDefault();
            e.stopPropagation();
            apply(value);
        }
        else if (e.key === "Escape") {
            e.preventDefault();
            e.stopPropagation();
            setValue(getValue());
            setError("");
            if (editor.onCancel)
                editor.onCancel();
        }
    };

    const handleBlur = (input: string) => {
        if (valueChanged.current) {
            if (!apply(input)) {
                setValue(getValue());
            }
        }
        setError("");
        if (editor.onCancel)
            editor.onCancel();
        valueChanged.current = false;
    };

    React.useEffect(() => setValue(getValue()), [editor.value]);
    React.useEffect(() => setError(""), [editor.value]);

    return (
        <TextCellEditorStyles>
            <Autocomplete
                suggestions={suggestions || []}
                value={value}
                onInputChange={onChange}
                onInputBlur={(_, input) => handleBlur(input)}
                onSuggestionSelect={apply}
                onInputKeyUp={handleKeyUp}
                autoFocus={autoFocus}
                inputStyle={{ margin: 0, outlineColor: error.length !== 0 ? "var(--theia-inputValidation-errorBackground)" : undefined }}
            />
            <ErrorOverlay error={error} />
        </TextCellEditorStyles>
    );
};

export interface NumberCellEditorProps extends CellEditorProps<number> {
    min?: number;
    max?: number;
    allowDecimal?: boolean;
    allowBlank?: boolean;
    suggestions?: number[];
}

export const NumberCellEditor: React.FC<NumberCellEditorProps> = (props) => {
    const schema = React.useMemo(() => {
        let schema = yup.number()
            .min(props.min === undefined ? -Number.MAX_VALUE : props.min)
            .max(props.max === undefined ? Number.MAX_VALUE : props.max);
        if (props.allowDecimal === undefined || !props.allowDecimal)
            schema = schema.integer();
        return schema;
    }, []);
    const editor: CellEditor<string> = {
        value: props.editor.value?.toString(),
        onApply: (value: string) => {
            if (value.length === 0 && props.allowBlank)
                value = "";
            else if (editor.validate!(value) !== undefined)
                return;
            if (props.editor.onApply)
                props.editor.onApply(schema.cast(value)!);
        },
        validate: (value: string) => {
            if (value.length === 0 && props.allowBlank)
                return undefined;
            let _value: number;
            try {
                _value = schema.validateSync(value)!;
            } catch (error) {
                return error.errors[0];
            }
            if (props.editor.validate)
                return props.editor.validate(_value);
        },
        readOnly: props.editor.readOnly
    };
    // @ts-ignore
    return <TextCellEditor {...props}
        editor={editor}
        suggestions={props.suggestions?.map(n => ({ label: n.toString() }))}
    />;
};

export const BooleanCellEditor: React.FC<CellEditorProps<boolean>> = ({
    editor
}) => {
    const getValue = () =>
        editor.value === undefined ? false : editor.value;
    const [value, setValue] = React.useState(getValue());
    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (editor.readOnly)
            return;
        setValue(e.target.checked);
        if (editor.onApply)
            editor.onApply(e.target.checked);
    };
    React.useEffect(() => setValue(getValue()), [editor.value]);

    return (
        <input
            type="checkbox"
            checked={value}
            onChange={onChange}
            className="theia-input"
            style={{ marginLeft: "5px" }}
        />
    );
};

export interface ComboCellEditorProps extends CellEditorProps<string> {
    values: ComboCellEditorProps.Option[];
}

export namespace ComboCellEditorProps {
    export interface Option {
        value: any;
        name: string;
    }
}

export const ComboCellEditor: React.FC<ComboCellEditorProps> = ({
    editor,
    values
}) => {
    const getValue = () =>
        editor.value === undefined ? "" : editor.value;
    const [value, setValue] = React.useState(getValue());
    React.useEffect(() => setValue(getValue()), [editor.value]);
    const handleChange = (
        e: React.ChangeEvent<HTMLSelectElement | HTMLInputElement>
    ) => {
        if (editor.readOnly)
            return;
        setValue(e.target.value);
        if (editor.onApply)
            editor.onApply(e.target.value);
    };

    return (
        <select
            value={value}
            onChange={handleChange}
            style={{ width: "100%", margin: 0, padding: 0 }}
            className="theia-input"
        >
            {values.map(value => (
                <option key={value.name} value={value.value}>
                    {value.name}
                </option>
            ))}
        </select>
    );
};

export interface DialogCellEditorProps<T> extends CellEditorProps<T> {
    openDialog: (value: T) => MaybePromise<T>;
    renderLabel?: (value: T) => React.ReactNode;
    tooltip?: string;
}

export const DialogCellEditor: React.FC<DialogCellEditorProps<any>> = ({
    editor,
    openDialog,
    renderLabel,
    tooltip
}) => {
    const getValue = () =>
        editor.value === undefined ? false : editor.value;
    const [value, setValue] = React.useState(getValue());
    React.useEffect(() => setValue(getValue()), [editor.value]);

    const handleClick = () => {
        openDialog(value)?.then((newValue: any) => {
            if (newValue !== undefined && !editor.readOnly) {
                setValue(newValue);
                if (editor.onApply)
                    editor.onApply(newValue);
            }
        });
    };

    const handeKeyDown = (e: React.KeyboardEvent) => {
        if (e.key === "Enter")
            handleClick();
    };

    return (
        <div
            style={{
                display: "grid",
                gridTemplateColumns: "1fr max-content",
                margin: 0,
                padding: "4px",
                width: "100%",
            }}
            title={tooltip || (value instanceof Array ? value.join("\n") : value.toString())}
            tabIndex={0}
            onKeyDown={handeKeyDown}
        >
            <div
                style={{
                    whiteSpace: "nowrap",
                    overflow: "hidden",
                    textOverflow: "ellipsis"
                }}
                onDoubleClick={handleClick}
            >
                {renderLabel && renderLabel(value)}
                {!renderLabel && (value instanceof Array ? value.join(", ") : value)}
            </div>
            <div onClick={handleClick} title={messages.click_to_edit} style={{ cursor: "pointer" }}>
                <div className="martini-tree-icon martini-edit-icon" />
            </div>
        </div>
    );
};


const ErrorOverlayStyles = styled.div`
    position: absolute;
    top: 24px;
    width: 100%;
    background-color: var(--theia-inputValidation-errorBackground);
    z-index: 10;
    text-overflow: ellipsis;
    overflow: hidden;

    .error-message {
        white-space: pre;
        height: 24px;
        font-size: 10px;
        padding: 4px;
        min-width: 0;
        text-overflow: ellipsis;
        overflow: hidden;
    }
`;

export const ErrorOverlay: React.FC<{ error: string; }> = ({ error }) => {
    if (!error || error.length === 0)
        return <></>;

    return <ErrorOverlayStyles>
        <div className="error-message" title={error}>{error.replace(/[\r\n]/g, " ")}</div>
    </ErrorOverlayStyles>;
};
