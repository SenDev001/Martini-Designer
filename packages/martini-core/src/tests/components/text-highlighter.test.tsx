import * as React from "react";
import * as Renderer from "react-test-renderer";
import { TextHighlighter } from "../../browser/components/text-highlighter";

test("Should highlight the text", () => {
    const component = Renderer.create(<TextHighlighter
        search="test"
    >
        This is a test.
    </TextHighlighter>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Should highlight the text in children elements", () => {
    const component = Renderer.create(<TextHighlighter
        search="null"
    >
        <div>
            <span>Response </span>
            <span>HttpMethods.http</span>
            <span>(</span>
            <span> </span>
            <span>request: </span>
            <span>null</span>
            <span>, </span>
            <span>auth: </span>
            <span>null</span>
            <span>, </span>
            <span>returnAs: </span>
            <span>"String"</span>
            <span>, </span>
            <span>responseType: </span>
            <span>"Auto Detect"</span>
            <span>, </span>
            <span>throwHttpExceptions: </span>
            <span>false</span>
            <span>, </span>
            <span>trackable: </span>
            <span>null</span>
            <span> </span>
            <span>)</span>
            <span> throws </span>
            <span>IOException</span>
        </div>
    </TextHighlighter>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

