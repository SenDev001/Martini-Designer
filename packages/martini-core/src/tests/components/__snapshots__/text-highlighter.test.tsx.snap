// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should highlight the text 1`] = `
<span
  className="thWrapper"
>
  <span
    className={null}
  >
    This is a 
  </span>
  <span
    className={null}
    style={
      Object {
        "backgroundColor": "var(--theia-list-filterMatchBackground)",
      }
    }
  >
    test
  </span>
  <span
    className={null}
  >
    .
  </span>
</span>
`;

exports[`Should highlight the text in children elements 1`] = `
<div>
  <span>
    <span
      className={null}
    >
      Response 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      HttpMethods.http
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      (
    </span>
  </span>
  <span>
    <span
      className={null}
    >
       
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      request: 
    </span>
  </span>
  <span>
    <span
      className={null}
      style={
        Object {
          "backgroundColor": "var(--theia-list-filterMatchBackground)",
        }
      }
    >
      null
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      , 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      auth: 
    </span>
  </span>
  <span>
    <span
      className={null}
      style={
        Object {
          "backgroundColor": "var(--theia-list-filterMatchBackground)",
        }
      }
    >
      null
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      , 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      returnAs: 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      "String"
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      , 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      responseType: 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      "Auto Detect"
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      , 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      throwHttpExceptions: 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      false
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      , 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      trackable: 
    </span>
  </span>
  <span>
    <span
      className={null}
      style={
        Object {
          "backgroundColor": "var(--theia-list-filterMatchBackground)",
        }
      }
    >
      null
    </span>
  </span>
  <span>
    <span
      className={null}
    >
       
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      )
    </span>
  </span>
  <span>
    <span
      className={null}
    >
       throws 
    </span>
  </span>
  <span>
    <span
      className={null}
    >
      IOException
    </span>
  </span>
</div>
`;
